### DISCUSSION TITLE
Generating an array of column identifiers (e.g. A -> AA)

### DISCUSSION THREAD

**Discussion by pwaring** [PHPOffice/PhpSpreadsheet#1802] [Q&A] [answered]
I have a spreadsheet where I want to auto-size all the columns I have populated, as well as the first cell in each column (bold for headings). At the moment I am using the built-in PHP `range` function to generate an array of column identifiers, then calling the relevant functions on those columns/cells:

```php
foreach(range('A', 'Z') as $column) {
    $worksheet->getCell("{$column}1")->getStyle()->getFont()->setBold(true);
    $worksheet->getColumnDimension($column)->setAutoSize(true);
}
```

That works fine when the range covers single-character cells, but `range('A', 'AA')` generates a single array with the character 'A'. Is there a way to generate an array of columns between two identifiers? I know I could use column numbers instead, but everything else in the code uses letters and they also make reading and debugging easier as I can compare them directly to the spreadsheet.

I'm currently using PhpSpreadsheet 1.8.2 (I know this is out of date but the underlying PHP version is 5.6 and I can't change that without upgrading the whole server).

**Comment by pwaring** [comment, marked as answer]
I eventually solved this via the use of `columnIndexFromString` and `getHighestColumn`:

```php
$highest_column_index = \PhpOffice\PhpSpreadsheet\Cell\Coordinate::columnIndexFromString($worksheet->getHighestColumn(1));

foreach (range(1, $highest_column_index) as $column) {
    $worksheet->getCellByColumnAndRow($column, 1)->getStyle()->getFont()->setBold(true);
    $worksheet->getColumnDimensionByColumn($column)->setAutoSize(true);
}
```

**Comment by MarkBaker** [comment]
You can write your own variant of `range()` for columns; that will follow perl/excel style, and isn't limited to A..Z:

I'm just using it as a closure here; but if it's something that you use frequently in your code then it can be written as a standalone function.

```php
$columnRange = function($startColumn, $endColumn) {
    ++$endColumn;
    for($column = $startColumn; $column !== $endColumn; ++$column) {
        yield $column;
    }
};

foreach($columnRange('A', 'AB') as $column) {
    echo $column, PHP_EOL;
}
```
